clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -1074,8 +1074,24 @@ public void onComplete(Map<TopicPartition, OffsetAndMetadata> offsets, Exception
 
         log.trace("Sending OffsetCommit request with {} to coordinator {}", offsets, coordinator);
 
-        return client.send(coordinator, builder)
+        heartbeat().sentHeartbeat(time.milliseconds());
```

Comment:
So I think there is a drawback to this. We rely on the heartbeat in order to detect when a rebalance is in progress. The coordinator will respond to the heartbeat with a REBALANCE_IN_PROGRESS error, which lets the consumer know that it should rejoin. However, we do not return this error in offset commit responses when the group is preparing a rebalance. In the worst case, if offset commits are sent more frequently than the heartbeat interval, then we may never detect a rebalance. Unfortunately, I'm not sure there's an easy way out of this problem without a protocol change. 

cc @guozhangwang Thoughts?

Reply:
Oh, I forgot about this, So we should alter the `GroupCoordinator` to response a `Errors.REBALANCE_IN_PROGRESS` when group is in `PreparingRebalance` state, for it has already return a `REBALANCE_IN_PROGRESS` when the group is in `CompletingRebalance`.

Reply:
Unfortunately, that would not be a compatible change. Older clients will not expect the REBALANCE_IN_PROGRESS error. The only way we could do it would be to bump the protocol version and return the error code only for the new version.

Reply:
Actually older clients does expect rebalance in progress error since it may still be returned in CompletingRebalance phase which is interpreted as a fatal CommitFailure (and in my recent PR: https://github.com/apache/kafka/pull/7312 we interpret it as a non-fatal error instead).

On the other hand, during PrepareRebalance phase the onPartitionsRevoked may likely commit offsets, so if we change the broker behavior to, say, "still successfully commit the offset but returns an error code" then old clients would interpret it as fatal errors.

So at the top of my head I feel that if we make this broker change, it is only safe to have 2.5 clients with PR 7312 to talk to a 2.5 broker that would still return error in PrepareRebalance phase and interpret it intelligently, but that breaks client-broker compatibility.. if we do not make this broker change, then we'll have to make client change more conservative, e.g. like "do not use two consecutive heartbeat response to update the heartbeat last time".

Reply:
Sorry for my late, It really takes several days to review the rebalance code, and here is my conclusion:
1.  `GroupCoordinator` may respond `REBALANCE_IN_PROGRESS` when receiving `CommitOffsetReq` but only when `CompletingRebalance`, so we should also respond when `PreparingRebalance`; 
2. client already has some not so good resolution on receiving `REBALANCE_IN_PROGRESS`, and it will be better after #7312 is merged.

my design is:
1. bump the protocol version of `CommitOffsetReq` and add a parameter to `GroupCoordinator.doCommitOffsets` to identify whether should we respond `REBALANCE_IN_PROGRESS`
2. client just act as what I have done in the PR.

and the interaction is as follows:
1. if an OLD client commits offset, it just acts as previous, and the broker will also act as previous
2. if a new client commits offset, it will update heartbeat, the broker may respond a  `REBALANCE_IN_PROGRESS`, the client also have the logic to handle it.


And by the way, I misoperated my git and all my repo were deleted, and again I deleted my GitHub Kafka repo, So now the pr is from an `unknown repository`, I need to close this PR and restart a new one, sincerely apologies!

Reply:
after I finished the coding, I find it's unnecessary to bump the version, how do you think about it